### ISSUE TITLE
Accessibility improvements

### ISSUE THREAD

**Issue by jardakotesovec** [asmecher/pflPlugin#4]
Latest PFL design comes with some accessibility challenges, starting this issue to discuss those.

First thing to focus on would be basic html structure of the label. 

<img width="407" alt="Screenshot 2024-04-10 at 11 43 25" src="https://github.com/asmecher/pflPlugin/assets/1590681/5b853732-554f-48ec-8cc9-499cd15b6290">

**Comment by jardakotesovec** [comment]
Initially I coded it as two column table. But now its even less table than before so I don't think that would work.

Instead from SR point of view I think about this as 'list of facts' might work better. And than the tricky part is just to how to combine the mark-up to keep the visual from above but also have more detailed fact listing.

Here is example what I have in mind. The struggle is that aria-hidden should not be used on focusable elements. Therefore I am somewhat dancing around the linkable parts and trying to fill in some additional information around it. Which might not work super well across languages. 

There would be also process to fine tune the fact statements that would be coming from SR, which John can probably help best, but first trying to get better understanding what strategy can we take.

Most experienced team member is Israel, will ask him for some additional ideas so we can work out the basic structure.

```html
<ul>
  <li aria-hidden="true">
    <span class="left-column">This article</span>
    <span class="right-column">Other articles</span>
  </li>
  <li>
    <span class="sr-only">This article has 2 peer reviewers while average for other articles is 2.3.</span>
    <span class="left-column" aria-hidden="true">Peer reviewers: 2</span>
    <span class="right-column" aria-hidden="true">2.4</span>
  </li>
  <li>
    <span class="left-column" aria-hidden="true">Reviewer</span>
    <span class="sr-only">You can check out Reviewer</span>
    <a href="">profiles</a></span>
  </li>
  <li aria-hidden="true">Author statements</li>
  <li>
    <span class="sr-only">Author statement on</span>
    <span>Data availability</span>
    <span><a href="">Yes</a></span>
  </li>
</ul>

```

**Comment by israelcefrin** [comment]
Hi all!!

Reviewing the information we are willing to present, I think we should consider manage it in small pieces of information or HTML components. If we follow the approach of a single  element ,`table` or a long unordered list `ul`, we may overwhelm users with too much info at once.

I've shared with Jarda an initial idea of structure, but I would like access to an HTML mockup to understand better how all the information is currently organized or structured. 
Please, check the [annotated image](https://www.figma.com/file/POzYgJhqelYSu9soPJdJc6/pfl-plugin-prototype?type=design&node-id=0%3A1&mode=design&t=aEC3Gdovqj9rCLry-1) that I've drafted.

The goal is to improve the semantic information while improving the accessibility for screen reader and keyboard navigation users. 
Also,  there is a possible Accessibility issue related to the  " Yes"  and the two/one letter links (component #7). The clickable area for each component on the screen should have a target size minimum of 24 CSS pixels - https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html

I haven't marked the elements below "author statement"(element #4) because I am unsure if it is a table related to the articles or a piece of separate information.  We have a "Yes" link and a number value on the same row. 

Maybe, with an HTML mockup, we will be able to address each element or component and add more context to development, e.g.:
1)  Heading level 2 |  screen reader announces: "Publication Facts, heading level 2" 
....
5) Table |  screen reader announces: "Journal comparison, table, 3 rows"

**Comment by jardakotesovec** [comment]
@israelcefrin Hi Israel, Alexandre posted updated version here  - https://github.com/asmecher/pflPlugin/pull/10

If you have moment to take it for spin. I will also have a more detailed look either later today or tomorrow.

**Comment by jardakotesovec** [comment]
@israelcefrin Just the testing. And there are some issues to address. Will ping you once thats improved, so you don't spend your time on same things and you can focus on additional improvements afterwards.

**Comment by jardakotesovec** [comment]
@israelcefrin 
Hi Israel,
I have done couple of adjustments on top of the Alexandre changes and now I am mostly happy SR experience, but would love to get your second opinion to know whether I am on good track and what still could be improved.

Styling is obviously all broken, focus is purely on SR experience. I am using voice over on mac with safari.

You can take it for spin - https://github.com/asmecher/pflPlugin/pull/11, my specific changes are in this commit - https://github.com/asmecher/pflPlugin/pull/11/commits/826deba16eed72f4ece3f2403bf20f9a5ec1ee2b

I will try to cover the adjustments I intentionally made:

### Tables done as 3 columns

<img width="312" alt="Screenshot 2024-05-30 at 13 05 08" src="https://github.com/asmecher/pflPlugin/assets/1590681/2df89382-8d81-426d-9ff3-d637fb6bcf79">

My interpretation of this design is that there are two headers - on top its 'this article' and 'other articles'. And on left its the individual statements. To be able to keep the design (which is not tabular) I opted to use the roles, not the html tags.

@israelcefrin This would be somewhat significant change in the existing mark up - so would like your confirmation whether this is good approach to take, before it would be adapted.  I applied it on two tables, the journal are still using the <table> tag, but it would make sense to follow the same structure there as well.
```
            <div role="table">
                <div role="row" class="header-row">
                    <div role="columnheader"><span class="pfl-sr-only">Author statement</span></div>
                    <div role="columnheader">{translate key="plugins.generic.pfl.thisArticle"}</div>
                    <div role="columnheader">{translate key="plugins.generic.pfl.otherArticles"}</div>
                </div>
                <div role="row">
                    <div role="rowheader">Data availability</div>
                    <div role="cell">No</div>
                    <div role="cell">16%</div>
                </div>
                <div role="row">
                    <div role="rowheader">External Funding</div>
                    <div role="cell">No</div>
                    <div role="cell">32%</div>
                </div>
                <div role="row">
                    <div role="rowheader">Competing interests</div>
                    <div role="cell">No</div>
                    <div role="cell">11%</div>
                </div>
            </div>

```

### Reduced aria-label use
aria-label is mainly for icons and things that needs some description. If things are well semanticaly structured SR will describe it well. For example in table SR will be quite informative about on what column/row person is and what are the headers. So for example using aria-label on td was causing duplicated announcement of 'this article' and 'other article'.


### Removed display:contents
This was causing in safari that the body was removed from accessibility tree. Not sure what intention was there with styling, but does not work accessibility wise.
```
publication-facts-label table tbody {
  display:contents;
}
```

### Added role="list"
Some browsers (safari in particular) [don't consider](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style#accessibility_concerns) <ul> being list when list-style is none. It requires to add the role explicitely to address that.

### Restructured the section 8 with Definition list
My understanding was that Israel was suggesting structure of 3 definitions (each of them consist of dt and dd):
```
            <dl>
                <dt class="pfl-list-item pfl-paragraph-item">
                    {translate key="plugins.generic.pfl.editorAndBoardMembers"}
                </dt>
                <dd>
                    <img src="{$baseUrl|concat:"/plugins/generic/pflPlugin/img/orcid.svg"}" alt="ORCiD logo image">

                    <a href="{$editorialTeamUrl}" target="_blank">{translate key="plugins.generic.pfl.profiles"}</a>

                </dd>


                <dt class="pfl-list-item">
                    <div class="pfl-indent">
                        {translate key="plugins.generic.pfl.academicSociety"}
                    </div>
                </dt>
                <dd>
                    <a href="" target="_blank">N/A</a>
                </dd>

                {if $pflPublisherUrl or $pflPublisherName} 
                    <dt class="pfl-last">
                        <div class="pfl-indent">
                                {translate key="plugins.generic.pfl.publisher"}

                        </div>
                    </dt>
                    <dd>
                        {if $pflPublisherUrl}
                            <a href="{$pflPublisherUrl|escape}" target="_blank">
                                {$pflPublisherName|escape|default:"&mdash;"}
                            </a>
                        {else}
                            {$pflPublisherName|escape|default:"&mdash;"}
                        {/if}
                    </dd>
                {/if}

            </dl>
```

### Made adjustment for 6 to make "Indexed In" as h3 and link to the list
This seems better fit for list to have the "Indexed in" as heading, instead of paragraph?

 

cc @DesignCodex

**Comment by israelcefrin** [comment]
Hi @jardakotesovec 

I will add my comments bellow:

> My interpretation of this design is that there are two headers - on top its Author statement, This article and Other articles. And on left its the individual statements. To be able to keep the design (which is not tabular) I opted to use the roles, not the html tags.

My only concern with this approach is using a SR , can we ensure that it will announce the full information for each data? E.g.:
`Author statement, This article,  ,Data availability, No` 
It will require some tweaking on using aria-description since we will not use HTML native table tags. Otherwise, SR will only announce the current element , isn't?  SR will announce all the table headers related to a cell before to announce the value of cell itself. It is possible to mimic this behavior with role attributes but you will also need to add aria attributes including the description. 


> Reduced aria label
> So, for example, using an aria-label on td was causing duplicated announcements of 'this article' and 'other article'.

You are correct, aria is intended to be used when the element or component can't provide a description or announce the value it contains. It is usually enough if you use semantic and natively correct HTML tags. 

> Added role="list"
>Some browsers (safari in particular) [don't consider](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style#accessibility_concerns) <ul> being list when list-style is none.

Yes, it is a workaround for browser with a lack of correct CSS support (I am looking at you Safari)

> My understanding was that Israel was suggesting structure of 3 definitions (each of them consist of dt and dd):

This would provide a number of items that SR would announce to the user at the beginning of the `<dl>`. Being one or 100 , the SR user can stop the reading whenever they want if it is too long or not interested on read all of them.


> This seems better fit for list to have the "Indexed in" as heading, instead of paragraph?

I agree!! Also, this an example which requires a workaround for the lack of description and the minimum target size on links. 
Using a SR it only announces "G S" for one of the links. We either can add a `aria-description` to the target of the URL , e.g. "Google Scholar" or adding a visually hidden `<span class="visuallyhidden"> google scholar</span>`. 

Target size WCAG definition: The size of the [target](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html#dfn-target) for [pointer inputs](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html#dfn-pointer-input) is at least 24 by 24 [CSS pixels](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html#dfn-css-pixel). 
Source: https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html

Other than that, I think the approach is great @jardakotesovec , testing with a SR is not always easy or fast. If you would like to test it in WindowsOS, I would recommend using NVDA, which is free and opensource solution. 
The stack: windows + chrome + NVDA is the most common combination among users, while Safari+ VoiceOver is the toolkit used the most by developers.

**Comment by jardakotesovec** [comment]
> > My interpretation of this design is that there are two headers - on top its Author statement, This article and Other articles. And on left its the individual statements. To be able to keep the design (which is not tabular) I opted to use the roles, not the html tags.
> 
> My only concern with this approach is using a SR , can we ensure that it will announce the full information for each data? E.g.: `Author statement, This article, ,Data availability, No` It will require some tweaking on using aria-description since we will not use HTML native table tags. Otherwise, SR will only announce the current element , isn't? SR will announce all the table headers related to a cell before to announce the value of cell itself. It is possible to mimic this behavior with role attributes but you will also need to add aria attributes including the description.

Is the concern about using the roles instead of the html tags? While testing with voiceover - it behave exactly the same as table. But maybe the support is bit worse across different combinations of SR and browsers compared to tags.. don't know.

You would recommend to stick with two columns solution to be able to follow the layout and stick with html tags?



> > My understanding was that Israel was suggesting structure of 3 definitions (each of them consist of dt and dd):
> 
> This would provide a number of items that SR would announce to the user at the beginning of the `<dl>`. Being one or 100 , the SR user can stop the reading whenever they want if it is too long or not interested on read all of them.
> 
Right, was just asking whether the html coding above is what you expected (having 3 pairs of dt&dd). Or is more fitting the coding that Alexandre provided, which is 1 dt + 2 dd.

**Comment by israelcefrin** [comment]
> Is the concern about using the roles instead of the html tags? While testing with voiceover - it behave exactly the same as table. 

If the whole information has been announced (table header value, row header value, cell value) then it is fine. I was running a local test and it seemed announcing only the last one.

> Right, was just asking whether the html coding above is what you expected (having 3 pairs of dt&dd). 

Ohhhh, ok. Yes, using H3 improves the readability too. 
Using headings tags with `IDs` is even better to share a direct link to it. See the ToC with mapping headings using this version:

![image](https://github.com/asmecher/pflPlugin/assets/4992902/125253f6-a62a-42cf-b10d-e1fd4cb7ad6a)

**Comment by jardakotesovec** [comment]
@israelcefrin Did test the role based table today with NVDA and Chrome and seems working correctly - it always tell me what column/row I am.. so its easy to navigate through it.

**Comment by jardakotesovec** [comment]
@DesignCodex After my testing all accessibility related recommendations are listed in https://github.com/asmecher/pflPlugin/issues/4#issuecomment-2139331029

And additionally important feedback from Israel regarding listing the indexes:
>Also, this an example which requires a workaround for the lack of description and the minimum target size on links.
Using a SR it only announces "G S" for one of the links. We either can add a aria-description to the target of the URL , e.g. "Google Scholar" or adding a visually hidden <span class="visuallyhidden"> google scholar</span>.

> Target size WCAG definition: The size of the [target](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html#dfn-target) for [pointer inputs](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html#dfn-pointer-input) is at least 24 by 24 [CSS pixels](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html#dfn-css-pixel).
Source: https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html

If any of these is unclear feel free to reach out either here or mattermost for faster response. Thank you!
